casper/src/main/resources/AsyncMap.rho

```diff
@@ -0,0 +1,208 @@
+//       AsyncMap!("set", map, "some key", "some val", *ret1) |
+//       AsyncMap!("set", map, "monkey", "some other val", *ret2) |
+//       for (_ <- ret1; _ <- ret2) {
+//         o!("Value set, getting") |
+//         AsyncMap!("get", map, "some key", *ret1) |
+//         AsyncMap!("fastUnsafeGet", map, "monkey", *ret2) |
+//         AsyncMap!("get", map, "some unused key", *ret3) |
+//         for (@val1 <- ret1; @val2 <- ret2; @val3 <- ret3) {
+//           // Should print ["Got these from the map: ", ["some val"], ["some other val"], []]
+//           o!(["Got these from the map: ", val1, val2, val3])
+//         }
+//       }
+//     }
+//   }
+// }
+
+new AsyncMap, o(`rho:io:stdout`) in {
+  new MakeCell, ByteArrayToNybbleList, AsyncMapSetter, AsyncMapGetter, powersCh in {
+    powersCh!([1,2,4,8,16,32,64,128,256,512,1024,2048,4096,8192,16384,32768,65536]) |
+    for (@powers <- powersCh) {
+      contract MakeCell(@initVal, cell) = {
+        new cellStore in {
+          cellStore!(initVal) |
+          contract cell(@"get", ret) = {
+            // TODO: convert to peek
+            for (@val <- cellStore) {
+              cellStore!(val) |
+              ret!(val)
+            }
+          } |
+          
+          contract cell(@"set", @newVal, ack) = {
+            for (_ <- cellStore) {
+              cellStore!(newVal) |
+              ack!(Nil)
+            }
+          } |
+          
+          contract cell(@"setBit", @bit, ack) = {
+            for (@val <- cellStore) {
+              // val | (1 << bit)
+              // Bitwise operators would be really nice to have!
+              cellStore!((val % powers.nth(bit)) +
+                (val / powers.nth(bit + 1)) * powers.nth(bit + 1) +
+                powers.nth(bit)) |
+              ack!(Nil)
+            }
+          } |
+          
+          contract cell(@"setMap", @newKey, @newVal, ack) = {
+            for (@val <- cellStore) {
+              ack!(Nil) |
+              if (val == 0) { cellStore!({newKey: newVal}) }
+              else { cellStore!(val.set(newKey, newVal)) }
+            }
+          }
+        }
+      } |
+
+      contract ByteArrayToNybbleList(@ba, @n, @len, @acc, ret) = {
+        if (n == len) {
+          ret!(acc)
+        } else {
+          ByteArrayToNybbleList!(ba, n+1, len, acc ++ [ ba.nth(n) % 16, ba.nth(n) / 16 ], *ret)
+        }
+      } |
+
+      contract AsyncMap(@"init", @depth, ret) = {
+        new map in {
+          MakeCell!(0, (*map, [])) |
+          @(*map, "depth")!!(depth) |
+          ret!(*map)
+        }
+      } |
+
+      contract AsyncMapGetter(@map, @nybList, @n, @len, @suffix, ret) = {
+        // Look up the value of the cell at [map, nybList.slice(0, n + 1)]
+        new valCh, restCh in {
+          @(map, nybList.slice(0, n))!("get", *valCh) |
+          for (@val <- valCh) {
+            if (n == len) {
+              restCh!(val.get(suffix)) |
+              for (@rest <- restCh) {
```

Comment:
we could use `match` for binding a variable without incurring a COMM.

Reply:
Actualy the if beneath could be absorbed by the match?

With 'storing entries under keys' though we should just be able to `ret!(val.get(key))` though

Reply:
Removed comm, merged match and if.  Can't just `ret!(val.get(key))` without changing the API to return `Nil` instead of `[]`.  Rholang's built-in maps can't distinguish between storing Nil at a key and having nothing stored at the key.  I chose to make the distinction in this map implementation, but if you prefer consistency, I can use the same API as the built-in map instead.

Reply:
Yeah, that's unfortunate. I pick consistency though, partly because List is confusing when used as a poor man's Option.

Reply:
Done